### DISCUSSION TITLE
Dynamically Expanding Column Chart with Snapping and Scroll Events

### DISCUSSION THREAD

**Discussion by asm4** [patrykandpatrick/vico#875] [Questions]
### Question

Hi Vico team,

I'm working on a bar chart using Vico and I'm trying to create a more customized scrolling experience. I have a couple of key features in mind:

1. **Configurable Visible Time Periods**: I want to control the visible time range on the chart, for example, showing one week or one month's worth of data at a time.
2. **Dynamically Expanding Column Layers**: The column widths should adjust dynamically to fill the available screen space, based on the visible time period. So, if a week is displayed, the 7 daily bars should expand to fill the width. On a narrower screen, the bars would be thinner but still use the full width.
3. **Scroll Snapping**: Ideally, I'd like the chart to "snap" to the next time period (e.g., the next week or month) when the user scrolls. This would provide a more deliberate and clear navigation experience.
4. **Scroll Event Handling**: To implement the snapping behavior, I need more control over the scroll state. It would be great to have callbacks or listeners that indicate when the user starts and stops scrolling. This would allow me to detect scroll events and programmatically adjust the chart's position to snap to the desired time period.

Is it currently possible to achieve this level of customization with Vico? If so, could you please provide some guidance, examples, or point me to relevant documentation?

If these features aren't currently supported, I'd like to request them as additions to the library. I believe they would significantly enhance the developer experience and flexibility of Vico charts, especially for time-series data visualization.

I took a shot at trying to make the bars expand dynamically and the best thing I could come up with was not ideal. I had to stick it into a `BoxWithConstraints` to get the `maxWidth` of the overall width of the entire graph (including the y-axis). From there, I tried to estimate the width of the y-axis and subtract that from the total width before dividing by the number of bars I wanted shown. [I have attached that for reference](https://gist.github.com/asm4/3026e4b83a46b9f4478ae83006b4e2c3).

Thanks for your time and consideration!

<img width="526" alt="Screenshot 2024-09-12 at 8 49 00 AM" src="https://github.com/user-attachments/assets/e2aa886a-aa50-42b5-b7c4-3c5b8c5ceaec">
<img width="317" alt="Screenshot 2024-09-12 at 8 48 10 AM" src="https://github.com/user-attachments/assets/4d028343-8f96-4c73-9670-349a226ba53f">
<img width="499" alt="Screenshot 2024-09-12 at 8 48 29 AM" src="https://github.com/user-attachments/assets/da2846f5-5356-4d11-ab4c-8206da218cae">
<img width="253" alt="Screenshot 2024-09-12 at 8 48 51 AM" src="https://github.com/user-attachments/assets/521fcba0-a7fb-4789-8cf4-4449cac7a738">

### Vico version(s)

2.0.0-alpha.28

### UI framework(s)

Jetpack Compose

**Comment by Gowsky** [comment]
Hello, @asm4. Points 1 and 2 are possible — see `rememberVicoZoomState` and `Zoom.x`. For example, if the _x_ values assigned to consecutive days are consecutive integers, you'd use `Zoom.x(7.0)` to show one week of data at a time. The seven columns would perfectly fill the viewport width, as desired.

Snapping is a planned addition. We'll also look into adding a property similar to `ScrollState.isScrollInProgress` to `VicoScrollState`.
